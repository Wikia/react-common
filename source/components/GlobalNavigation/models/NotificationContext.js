import React from 'react';

const NotificationsContext = React.createContext({});

export default NotificationsContext;

export const NotificationsProvider = NotificationsContext.Provider;
export const NotificationsConsumer = NotificationsContext.Consumer;
