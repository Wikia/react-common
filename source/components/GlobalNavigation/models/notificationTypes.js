export const notificationTypes = {
    discussionUpvotePost: 'discussion-upvote-post',
    discussionUpvoteReply: 'discussion-upvote-reply',
    discussionReply: 'discussion-reply',
    announcement: 'announcement',
};

export const isDiscussionReply = type => type === notificationTypes.discussionReply;
export const isDiscussionReplyUpvote = type => type === notificationTypes.discussionUpvoteReply;
export const isDiscussionPostUpvote = type => type === notificationTypes.discussionUpvotePost;
export const isAnnouncement = type => type === notificationTypes.announcement;
