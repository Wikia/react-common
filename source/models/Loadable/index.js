// eslint-disable-next-line no-restricted-imports
import Model from '../Model';

export const LOADABLE_STATES = Object.freeze({
    UNINITIALIZED: 'UNINITIALIZED',
    IN_PROGRESS: 'IN_PROGRESS',
    COMPLETED: 'COMPLETED',
    ERROR: 'ERROR',
});
const LOADABLE_STATES_VALUES = Object.values(LOADABLE_STATES);

const LOADABLE_MODEL_KEYS = Object.freeze({
    state: '_loadable_state',
});

const loadableSchema = {
    [LOADABLE_MODEL_KEYS.state]: LOADABLE_STATES.UNINITIALIZED,
};

const LoadableModel = (defaultValues, name) => class extends Model({
    ...defaultValues,
    ...loadableSchema,
}, name) {
    _getLoadingState() {
        return this.get(LOADABLE_MODEL_KEYS.state);
    }

    _withLoadingState(state) {
        if (LOADABLE_STATES_VALUES.includes(state)) {
            return this.set(LOADABLE_MODEL_KEYS.state, state);
        }

        console.error(`LoadableModel._withLoadingState called with incorrect param: "${state}". Allowed states: ${LOADABLE_STATES_VALUES.join(', ')}.`);

        return this;
    }

    isLoadingUninitialized() {
        return this._getLoadingState() === LOADABLE_STATES.UNINITIALIZED;
    }

    isLoadingInProgress() {
        return this._getLoadingState() === LOADABLE_STATES.IN_PROGRESS;
    }

    isLoadingCompleted() {
        return this._getLoadingState() === LOADABLE_STATES.COMPLETED;
    }

    isLoadingError() {
        return this._getLoadingState() === LOADABLE_STATES.ERROR;
    }

    /**
     * is loading completed or error?
     */
    isLoadingFinished() {
        return this.isLoadingCompleted() || this.isLoadingError();
    }

    withLoadingInProgress() {
        return this._withLoadingState(LOADABLE_STATES.IN_PROGRESS);
    }

    withLoadingCompleted() {
        return this._withLoadingState(LOADABLE_STATES.COMPLETED);
    }

    withLoadingError() {
        return this._withLoadingState(LOADABLE_STATES.ERROR);
    }
};

export default LoadableModel;
