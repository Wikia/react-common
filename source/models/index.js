/**
 * Import test - please put all the components here
 * so Jest can verify that all the components are
 * importable.
 */

// eslint-disable-next-line
export { default as Model } from './Model';
