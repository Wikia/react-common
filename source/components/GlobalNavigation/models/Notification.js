import merge from 'lodash/merge';
import get from 'lodash/get';

import { convertToTimestamp } from '../utils/isoTime';

import { notificationTypes } from './notificationTypes';

const defaultProps = {
    title: null,
    snippet: null,
    timestamp: null,
    communityName: null,
    type: null,
    isUnread: false,
    totalUniqueActors: 1,
    latestActors: null,
    uri: null,
};

function mapData(notificationData) {
    return {
        title: get(notificationData, 'refersTo.title'),
        snippet: get(notificationData, 'refersTo.snippet'),
        uri: get(notificationData, 'refersTo.uri'),
        latestEventUri: get(notificationData, 'events.latestEvent.uri'),
        timestamp: convertToTimestamp(get(notificationData, 'events.latestEvent.when')),
        communityName: get(notificationData, 'community.name'),
        communityId: get(notificationData, 'community.id'),
        isUnread: notificationData.read === false,
        totalUniqueActors: get(notificationData, 'events.totalUniqueActors'),
        latestActors: createActors(get(notificationData, 'events.latestActors')),
        type: getNotificationType(notificationData),
    };
}

function createActors(actors) {
    return actors.map(actor => merge({}, actor, { profileUrl: getProfileUrl(actor.username) }));
}

function getProfileUrl(name) {
    return `${window.location.origin}/wiki/User:${name}`;
}

function getNotificationType(apiData) {
    if (apiData.type === 'upvote-notification') {
        if (apiData.refersTo.type === 'discussion-post') {
            return notificationTypes.discussionUpvoteReply;
        }

        return notificationTypes.discussionUpvotePost;
    }

    if (apiData.type === 'replies-notification') {
        return notificationTypes.discussionReply;
    }

    if (apiData.type === 'announcement-notification') {
        return notificationTypes.announcement;
    }

    return null;
}

class Notification {
    constructor(data) {
        Object.keys(data).forEach((key) => {
            this[key] = data[key];
        });
    }

    static build(notificationData) {
        return new this(merge({}, defaultProps, mapData(notificationData)));
    }
}

export default Notification;
