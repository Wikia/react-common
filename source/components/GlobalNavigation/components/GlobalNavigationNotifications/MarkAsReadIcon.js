import React, { useCallback, useContext } from 'react';
import PropTypes from 'prop-types';

import Icon from '../../../Icon';

import NotificationsContext from '../../utils/NotificationContext';
import { isAnnouncement, isDiscussionReply } from '../../models/notificationTypes';

function getIconName(type) {
    if (isDiscussionReply(type)) {
        return 'comment-small';
    }

    if (isAnnouncement(type)) {
        return 'flag';
    }
    return 'upvote-small';
}

const MarkAsReadIcon = ({ track, model }) => {
    const { type, isUnread, uri } = model;
    const { markAsRead } = useContext(NotificationsContext);
    const onClick = useCallback((event) => {
        event.stopPropagation();

        if (!isUnread) {
            return;
        }

        track({
            action: 'click',
            category: 'on-site-notifications',
            label: `mark-as-read-${type}`,
        });

        markAsRead(uri);
    }, [markAsRead, track, model]);

    return (
        <div className="wds-notification-card__icon-wrapper" onClick={onClick}>
            <Icon name={getIconName(type)} />
        </div>
    );
};

MarkAsReadIcon.propTypes = {
    // eslint-disable-next-line react/forbid-prop-types
    model: PropTypes.object.isRequired,
    track: PropTypes.func.isRequired,
};

export default MarkAsReadIcon;
