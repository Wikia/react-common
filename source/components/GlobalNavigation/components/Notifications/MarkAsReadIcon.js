import React, { useCallback, useContext } from 'react';
import PropTypes from 'prop-types';

import Icon from '../../../Icon';

import NotificationsContext from '../../utils/NotificationContext';
import { isAnnouncement, isDiscussionReply } from '../../models/notificationTypes';

function getIconName(type) {
    if (isDiscussionReply(type)) {
        return 'comment-small';
    }

    if (isAnnouncement(type)) {
        return 'flag-small';
    }

    return 'upvote-small';
}

const MarkAsReadIcon = ({ track, model }) => {
    const { type, isUnread, uri } = model;
    const { markAsRead } = useContext(NotificationsContext);
    const onClick = useCallback((event) => {
        event.stopPropagation();
        event.preventDefault();

        if (!isUnread) {
            return;
        }

        track({
            action: 'click',
            category: 'on-site-notifications',
            label: `mark-as-read-${type}`,
        });

        markAsRead(uri);
    }, [markAsRead, track, model]);

    return (
        <div className="wds-notification-card__icon-wrapper" onClick={onClick} role="presentation">
            <Icon name={getIconName(type)} />
        </div>
    );
};

MarkAsReadIcon.propTypes = {
    model: PropTypes.object.isRequired,
    track: PropTypes.func.isRequired,
};

export default MarkAsReadIcon;
