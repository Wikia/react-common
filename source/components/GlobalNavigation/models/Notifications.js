import get from 'lodash/get';

import { convertToIsoString } from '../utils/isoTime';
import Notification from './Notification';

function getNextPageLink(data) {
    return get(data, '_links.next') || null;
}

class Notifications {
    unreadCount = 0;
    nextPage = 0;
    notifications = [];

    constructor(apiInstance) {
        this.api = apiInstance;
    }

    static build(apiInstance) {
        return new this(apiInstance);
    }

    loadFirstPage() {
        return this.loadPage('/notifications?contentType=discussion-upvote&contentType=discussion-post&contentType=announcement-target');
    }

    loadPage(pageLink) {
        return this.api
            .fetchFromOnSiteNotifications(pageLink)
            .then(response => {
                this.notifications = this.mapNotifications(response.notifications);
                this.unreadCount = this.getUpdatedUnreadCount();

                return getNextPageLink(response);
            });
    }

    markAsRead(notification, willUnloadPage) {
        if (!notification.isUnread) {
            return Promise.resolve();
        }

        return notification.markAsRead(willUnloadPage)
            .then(() => {
                this.unreadCount = this.getUpdatedUnreadCount();
            });
    }

    markAllAsRead() {
        const since = convertToIsoString(this.notifications[0].timestamp);

        return this.api.fetchFromOnSiteNotifications(`/notifications/mark-all-as-read`, {
                body: JSON.stringify({ since }),
                headers: {
                    'Content-Type': 'application/json'
                },
                method: 'POST',
            }).then(() => {
                this.notifications = () => {}; // todo
                this.unreadCount = 0;
            });
    },

    mapNotifications(notifications) {
        return (notifications || []).map(notification => Notification.build(notification, this.api));
    }

    getUpdatedUnreadCount() {
        return this.notifications.filter(notification => notification.isUnread).length;
    }
}

