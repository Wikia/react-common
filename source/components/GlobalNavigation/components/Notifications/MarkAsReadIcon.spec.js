import React, { useContext, useCallback } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme/build';
import merge from 'lodash/merge';

import { notificationTypes } from '../../models/notificationTypes';

import MarkAsReadIcon from './MarkAsReadIcon';

const defaultProps = {
    model: {
        type: notificationTypes.discussionReply,
        isUnread: true,
        uri: 'some://valid.url',
    },
    track: () => null,
};

const mockDefaultContext = { markAsRead: () => null };

function renderComponent(props) {
    const renderer = new ShallowRenderer();
    const computedProps = merge({}, defaultProps, props);

    renderer.render(<MarkAsReadIcon {...computedProps} />);

    return renderer.getRenderOutput();
}

// TODO: remove mocks when enzyme/react-test-renderer will fully support hooks
jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useContext: jest.fn(() => mockDefaultContext),
    useCallback: jest.fn(),
}));

/* eslint-disable-next-line jest/no-hooks */
beforeEach(() => {
    jest.resetModules();
});

test('MarkAsReadIcon renders correctly for discussion-reply', () => {
    expect(renderComponent()).toMatchSnapshot();
});

test('MarkAsReadIcon renders correctly for announcement', () => {
    expect(renderComponent({ model: { type: notificationTypes.announcement } })).toMatchSnapshot();
});

test('MarkAsReadIcon renders correctly for other types', () => {
    expect(renderComponent({ model: { type: notificationTypes.discussionUpvoteReply } })).toMatchSnapshot();
});

test('MarkAsReadIcon renders correctly for article-comment-reply', () => {
    expect(renderComponent({ model: { type: notificationTypes.articleCommentReply } })).toMatchSnapshot();
});

test('MarkAsReadIcon renders correctly for article-comment-at-mention', () => {
    expect(renderComponent({ model: { type: notificationTypes.articleCommentAtMention } })).toMatchSnapshot();
});

test('MarkAsReadIcon renders correctly for article-comment-reply-at-mention', () => {
    expect(renderComponent({ model: { type: notificationTypes.articleCommentReplyAtMention } })).toMatchSnapshot();
});

test('MarkAsReadIcon renders correctly for talk-page-message', () => {
    expect(renderComponent({ model: { type: notificationTypes.talkPageMessage } })).toMatchSnapshot();
});

test('MarkAsReadIcon renders correctly for message-wall-thread', () => {
    expect(renderComponent({ model: { type: notificationTypes.messageWallThread } })).toMatchSnapshot();
});

test('MarkAsReadIcon renders correctly for message-wall-post', () => {
    expect(renderComponent({ model: { type: notificationTypes.messageWallPost } })).toMatchSnapshot();
});

test('MarkAsReadIcon calls track and markAsRead on click when isUnread is true', () => {
    const markAsReadMock = jest.fn();
    const trackMock = jest.fn();
    const uriMock = 'some.url';
    const eventMock = { preventDefault: jest.fn(), stopPropagation: jest.fn() };
    useContext.mockImplementation(() => ({ ...mockDefaultContext, markAsRead: markAsReadMock }));
    useCallback.mockImplementation(callback => event => callback(event));
    const output = renderComponent({ model: { uri: uriMock }, track: trackMock });
    const wrapper = shallow(<div>{output}</div>);

    wrapper.find('.wds-notification-card__icon-wrapper').props().onClick(eventMock);

    expect(eventMock.preventDefault).toBeCalledWith();
    expect(eventMock.stopPropagation).toBeCalledWith();
    expect(markAsReadMock).toBeCalledWith(uriMock);
    expect(trackMock).toBeCalledWith(expect.any(Object));
});

test('MarkAsReadIcon does not call track and markAsRead on click when isUnread is false', () => {
    const markAsReadMock = jest.fn();
    const trackMock = jest.fn();
    const uriMock = 'some.url';
    const eventMock = { preventDefault: jest.fn(), stopPropagation: jest.fn() };
    useContext.mockImplementation(() => ({ ...mockDefaultContext, markAsRead: markAsReadMock }));
    useCallback.mockImplementation(callback => event => callback(event));
    const output = renderComponent({ model: { uri: uriMock, isUnread: false }, track: trackMock });
    const wrapper = shallow(<div>{output}</div>);

    wrapper.find('.wds-notification-card__icon-wrapper').props().onClick(eventMock);

    expect(eventMock.preventDefault).toBeCalledWith();
    expect(eventMock.stopPropagation).toBeCalledWith();
    expect(markAsReadMock).not.toBeCalled();
    expect(trackMock).not.toBeCalled();
});
