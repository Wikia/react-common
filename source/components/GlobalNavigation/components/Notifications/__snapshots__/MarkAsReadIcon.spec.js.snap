// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkAsReadIcon renders correctly for announcement 1`] = `
<div
  className="wds-notification-card__icon-wrapper"
  role="presentation"
>
  <Icon
    className=""
    name="flag"
    small={false}
    tiny={false}
  />
</div>
`;

exports[`MarkAsReadIcon renders correctly for discussion-reply 1`] = `
<div
  className="wds-notification-card__icon-wrapper"
  role="presentation"
>
  <Icon
    className=""
    name="comment-small"
    small={false}
    tiny={false}
  />
</div>
`;

exports[`MarkAsReadIcon renders correctly for other types 1`] = `
<div
  className="wds-notification-card__icon-wrapper"
  role="presentation"
>
  <Icon
    className=""
    name="upvote-small"
    small={false}
    tiny={false}
  />
</div>
`;
